File: packages/web/client/liminalMarkdownUtils.ts
import type { ReactNode } from "react";

const HTML_EMBED_LANGS = new Set(["html", "htm", "xhtml"]);

/** HTML void elements — no closing tag. */
const HTML_VOID = new Set([
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr",
]);

export function isHtmlEmbedLang(lang: string | undefined): boolean {
  return lang != null && HTML_EMBED_LANGS.has(lang.toLowerCase());
}

export function extractFencedCodeText(children: ReactNode): string {
  return String(children).replace(/\n$/, "");
}

export type StreamingHtmlFenceSplit = {
  /** Markdown for ReactMarkdown (text before an open ```html fence, or full text when closed). */
  outerMarkdown: string;
  /**
   * Body of the last ```html fence when it has no closing ``` yet.
   * `null` when no open fence or the fence is already closed (remark handles it).
   */
  htmlLive: string | null;
};

const HTML_FENCE_OPEN_RE = /```(?:html|htm|xhtml)\s*\r?\n/gi;

/**
 * While the model streams inside an unclosed ```html fence, remark does not emit a
 * `language-html` code node — the UI sees nothing until the closing fence arrives.
 * Peel off the live fence body so HtmlEmbedBlock can paint on every token.
 */
export function extractStreamingHtmlFence(text: string): StreamingHtmlFenceSplit {
  let lastOpenIdx = -1;
  let lastOpenLen = 0;
  let match: RegExpExecArray | null;
  HTML_FENCE_OPEN_RE.lastIndex = 0;
  while ((match = HTML_FENCE_OPEN_RE.exec(text)) !== null) {
    lastOpenIdx = match.index;
    lastOpenLen = match[0].length;
  }
  if (lastOpenIdx < 0) {
    return { outerMarkdown: text, htmlLive: null };
  }

  const afterOpen = text.slice(lastOpenIdx + lastOpenLen);
  const closeIdx = afterOpen.indexOf("```");
  if (closeIdx >= 0) {
    return { outerMarkdown: text, htmlLive: null };
  }

  return {
    outerMarkdown: text.slice(0, lastOpenIdx).trimEnd(),
    htmlLive: afterOpen,
  };
}

/**
 * Best-effort HTML for in-progress ```html fences: drop a truncated open tag at the
 * tail, then auto-close still-open elements so the browser can paint partial UI.
 */
export function balanceHtmlForStreamingPreview(html: string): string {
  let s = html.trim();
  if (!s) return "";

  const lastGt = s.lastIndexOf(">");
  const lastLt = s.lastIndexOf("<");
  if (lastLt > lastGt) {
    s = s.slice(0, lastLt).trimEnd();
  }
  if (!s) return "";

  const stack: string[] = [];
  const tagRe = /<\/?([a-zA-Z][\w:-]*)((?:\s+[^>]*)?)\s*(\/)?>/g;
  let match: RegExpExecArray | null;
  while ((match = tagRe.exec(s)) !== null) {
    const full = match[0];
    const name = match[1]!.toLowerCase();
    const selfClose = match[3] === "/" || full.endsWith("/>");
    if (selfClose || HTML_VOID.has(name)) continue;
    if (full.startsWith("</")) {
      const idx = stack.lastIndexOf(name);
      if (idx >= 0) stack.splice(idx);
    } else {
      stack.push(name);
    }
  }

  let out = s;
  for (let i = stack.length - 1; i >= 0; i--) {
    out += `</${stack[i]}>`;
  }
  return out;
}

File: apps/liminal_desktop/lib/ui/rich_message/liminal_markdown_utils.dart
// Ports packages/web/client/liminalMarkdownUtils.ts
library;

const _htmlEmbedLangs = {'html', 'htm', 'xhtml'};

const _htmlVoid = {
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr',
};

bool isHtmlEmbedLang(String? lang) =>
    lang != null && _htmlEmbedLangs.contains(lang.toLowerCase());

String extractFencedCodeText(String raw) =>
    raw.replaceAll(RegExp(r'\n$'), '');

class StreamingHtmlFenceSplit {
  const StreamingHtmlFenceSplit({
    required this.outerMarkdown,
    required this.htmlLive,
  });

  final String outerMarkdown;
  final String? htmlLive;
}

final _htmlFenceOpenRe = RegExp(
  r'```(?:html|htm|xhtml)\s*\r?\n',
  caseSensitive: false,
);

StreamingHtmlFenceSplit extractStreamingHtmlFence(String text) {
  var lastOpenIdx = -1;
  var lastOpenLen = 0;
  for (final match in _htmlFenceOpenRe.allMatches(text)) {
    lastOpenIdx = match.start;
    lastOpenLen = match.end - match.start;
  }
  if (lastOpenIdx < 0) {
    return StreamingHtmlFenceSplit(outerMarkdown: text, htmlLive: null);
  }

  final afterOpen = text.substring(lastOpenIdx + lastOpenLen);
  final closeIdx = afterOpen.indexOf('```');
  if (closeIdx >= 0) {
    return StreamingHtmlFenceSplit(outerMarkdown: text, htmlLive: null);
  }

  return StreamingHtmlFenceSplit(
    outerMarkdown: text.substring(0, lastOpenIdx).trimRight(),
    htmlLive: afterOpen,
  );
}

String balanceHtmlForStreamingPreview(String html) {
  var s = html.trim();
  if (s.isEmpty) return '';

  final lastGt = s.lastIndexOf('>');
  final lastLt = s.lastIndexOf('<');
  if (lastLt > lastGt) {
    final trimmed = s.substring(0, lastLt).trimRight();
    s = trimmed.isEmpty ? html.trim() : trimmed;
  }
  if (s.isEmpty) return '';

  final stack = <String>[];
  final tagRe = RegExp(r'<\/?([a-zA-Z][\w:-]*)((?:\s+[^>]*)?)\s*(\/)?>');
  for (final match in tagRe.allMatches(s)) {
    final full = match.group(0)!;
    final name = match.group(1)!.toLowerCase();
    final selfClose = match.group(3) == '/' || full.endsWith('/>');
    if (selfClose || _htmlVoid.contains(name)) continue;
    if (full.startsWith('</')) {
      final idx = stack.lastIndexOf(name);
      if (idx >= 0) stack.removeRange(idx, stack.length);
    } else {
      stack.add(name);
    }
  }

  var out = s;
  for (var i = stack.length - 1; i >= 0; i--) {
    out += '</${stack[i]}>';
  }
  return out;
}

final _closedHtmlFenceRe = RegExp(
  r'```(?:html|htm|xhtml)\s*\r?\n([\s\S]*?)```',
  caseSensitive: false,
);

enum MessageSegmentKind { markdown, htmlEmbed }

class MessageSegment {
  const MessageSegment({required this.kind, required this.content});

  final MessageSegmentKind kind;
  final String content;
}

/// Split assistant text into markdown vs closed ```html``` blocks.
List<MessageSegment> splitMessageSegments(String text) {
  final segments = <MessageSegment>[];
  var cursor = 0;
  for (final match in _closedHtmlFenceRe.allMatches(text)) {
    if (match.start > cursor) {
      final md = text.substring(cursor, match.start).trim();
      if (md.isNotEmpty) {
        segments.add(MessageSegment(kind: MessageSegmentKind.markdown, content: md));
      }
    }
    final html = match.group(1)?.trim() ?? '';
    if (html.isNotEmpty) {
      segments.add(MessageSegment(kind: MessageSegmentKind.htmlEmbed, content: html));
    }
    cursor = match.end;
  }
  if (cursor < text.length) {
    final tail = text.substring(cursor).trim();
    if (tail.isNotEmpty) {
      segments.add(MessageSegment(kind: MessageSegmentKind.markdown, content: tail));
    }
  }
  if (segments.isEmpty && text.trim().isNotEmpty) {
    segments.add(MessageSegment(kind: MessageSegmentKind.markdown, content: text));
  }
  return segments;
}

bool looksLikeInlineHtmlFragment(String text) {
  final t = text.trimLeft();
  if (!t.startsWith('<')) return false;
  return RegExp(
    r'^<(div|section|article|table|p|span|h[1-6]|ul|ol|header|footer|main)\b',
    caseSensitive: false,
  ).hasMatch(t);
}
